fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -800,57 +833,121 @@ public CompletableFuture<RecordCursorContinuation> rebalancePartitions(RecordCur
-                                                                                 RepartitioningLogMessages logMessages) {
+    public CompletableFuture<Integer> processPartitionRebalancing(@Nonnull final Tuple groupingKey,
+                                                                  int repartitionDocumentCount,
+                                                                  RepartitioningLogMessages logMessages) {
         if (repartitionDocumentCount <= 0) {
             throw new IllegalArgumentException("number of documents to move can't be zero");
         }
         return getAllPartitionMetaInfo(groupingKey).thenCompose(partitionInfos -> {
             // need to track the next partition id to use when creating a new one
-            int maxPartitionId = partitionInfos.stream().map(LucenePartitionInfoProto.LucenePartitionInfo::getId).max(Integer::compare).orElse(0);
+            RepartitioningContext repartitioningContext = new RepartitioningContext();
+            repartitioningContext.maxPartitionId = partitionInfos.stream().map(LucenePartitionInfoProto.LucenePartitionInfo::getId).max(Integer::compare).orElse(0);
             if (LOGGER.isTraceEnabled()) {
                 LOGGER.trace(partitionInfos.stream()
                         .sorted(Comparator.comparing(pi -> Tuple.fromBytes(pi.getFrom().toByteArray())))
                         .map(pi -> "pi[" + pi.getId() + "]@" + pi.getCount() + Tuple.fromBytes(pi.getFrom().toByteArray()) + "->" + Tuple.fromBytes(pi.getTo().toByteArray()))
                         .collect(Collectors.joining(", ", "Rebalancing partitions (group=" + groupingKey + "): ", "")));
             }
 
-            for (LucenePartitionInfoProto.LucenePartitionInfo partitionInfo : partitionInfos) {
-                if (partitionInfo.getCount() > indexPartitionHighWatermark) {
-                    // process one partition
-                    if (LOGGER.isDebugEnabled()) {
-                        LOGGER.debug(repartitionLogMessage("Repartitioning records", groupingKey, repartitionDocumentCount, partitionInfo));
+            for (int i = 0; i < partitionInfos.size(); i++) {
+                LucenePartitionInfoProto.LucenePartitionInfo partitionInfo = partitionInfos.get(i);
+
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(repartitionLogMessage("Repartitioning records", groupingKey, repartitionDocumentCount, partitionInfo));
+                }
+
+                final int currentPartitionCount = partitionInfo.getCount();
+
+                if (currentPartitionCount >= indexPartitionLowWatermark && currentPartitionCount <= indexPartitionHighWatermark) {
+                    // repartitioning not needed
+                    continue;
+                }
+
+                if (currentPartitionCount < indexPartitionLowWatermark) {
+                    // partitionInfos are ordered next -> previous
```

Comment:
`next` and `previous` I think are new terms only used here, perhaps replacing their usage with `newer` and `older` respectively.